docs/src/index.md

```diff
@@ -6,4 +6,356 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+    return y, pushforward
+end
+```
+again `y=foo(args, kwargs...)`,
+and _pushforward_ is a function to propagate the derivative information forwards at that point (more later).
+
+Almost always the _pushforward_/_pullback_ will be declared locally with-in the `ffrule`/`rrule`, and will be a _closure_ over some of the other arguments.
+
+### The propagators: pushforward and pullback
+
+!!! Terminology "Pushforward and Pullback"
+
+    _Pushforward_ and _Pullback_ are fancy words that the autodiff community recently stole from Differential Geometry.
+    The are broadly in agreement with the use of these terms in differential geometry. But any geometer will tell you these are the super-boring flat cases. Some will also frown at you.
+    Other terms that may be used include for _pullback_ the **backpropagator**, and by analogy for _pushforward_ the **forwardpropagator**, thus these are the _propagators_.
+    These are also good names because effectively they propagate wibbles and wobbles through them, via the chainrule.
+    (the term **backpropagator** may originate with ["Lambda The Ultimate Backpropagator"](http://www-bcl.cs.may.ie/~barak/papers/toplas-reverse.pdf) by Bearlmutter and Siskind, 2008)
+
+
+#### Core Important Idea:
+ - The **Pushforward** takes a wiggle in the _input space_, and tells what wobble you would create in the output space, by passing it through the function.
+ - The **Pullback** takes a wobble in the _output space_, and tells you what wiggle you would need to make in the _input_ space to achieve it.
+
+#### The anatomy of pushforward and pullback
+
+For our function `foo(args...; kwargs) = Y`:
+
+The pushforward is a function:
+```julia
+function pushforward(Δself, Δargs...)
+    ...
+    return ∂Y
+end
+```
+Note that there is one `Δargs...` per `arg` to the orginal function, and they are similar in type/structure to the ccorresponding inputs.
+Plus the `Δself` (don't worry we will be back to explain this soon).
+The `∂Y` will be similar in type/structure to the original function's output `Y`.
+In particular if that function returned a tuple then `∂Y` will be a tuple of same size.
+
+The input to the pushforward is often called the _pertubation_.
+If the function is `y=f(x)` often the pushforward will be written `ẏ=pushforward(ẋ)`.
+
+
+The pullback is a function
+```julia
+function pullback(ΔY)
+    ...
+    return ∂self, ∂args...
+end
+```
+
+Note that the pullback returns one `∂arg` per original `arg` to the function, plus one for the s
+
+The input to the pullback is often called the _seed_.
+If the function is `y=f(x)` often the pullback will be written `ȳ=pullback(x̄)`.
+
+
+!!! Terminology:
+    Sometimes _pertubation_, _seed_, _sensitivity_ will be used interchangeably, depending on task/subfield (_sensitivity_ analysis and perturbation analysis are apparently very big on just calling everying _sensitivity_ or _pertubation_ respectively.)
+    At the end of the day they are all _wibbles_ or _wobbles_.
```

Comment:
```suggestion
At the end of the day they are all _wibbles_ or _wobbles_.
```

Can you bring this up a line to make the source look more clean?

Reply:
This is actually a julia documentor specific markdown extension,
that will render ad a pretty block.

I think it needs to be indented (not 100% sure though)

https://docs.julialang.org/en/v1/stdlib/Markdown/#Markdown-Syntax-Extensions-1

Reply:
I checked with someone who knows and yes the indent is required.
(too bad code it hurts  github's renderer)